File: server/world/mapdata.py
"""小镇地图数据（服务端视图）：读取前后端共享的 town_map_v3.json。

设计说明（为什么这样设计）：
- 2026-08-26 起地图换 v3（the_ville 裁剪版，122×35）：转换管线
  client/scripts/the_ville_src/convert_ville_map.py 产出 town_map_v3.json
  （Tiled 标准字段 + 项目字段：walkable/sectors/locations/出生点）。
  改窗口/改地名/改站位 = 改该脚本重跑，无第二处维护。
- 服务端要校验移动合法性（寻路不穿墙），必须与前端渲染用同一份数据，
  所以共读一个 JSON：可行走网格由 maze.json 原作者标注的碰撞权威值
  预计算，裁剪边缘断开的孤岛已封死。
- MAP_DATA 同时暴露完整 dict：locations.py 从中派生站位点，地名白名单
  （planner）再从 locations 派生——地图数据是唯一源头。
- nearest_walkable：换图后旧坐标投射，防居民卡死在阻挡格。
"""

import json
from pathlib import Path

MAP_JSON = (
    Path(__file__).resolve().parents[2]
    / "client"
    / "public"
    / "assets"
    / "town_map_v3.json"
)

try:
    MAP_DATA: dict = json.loads(MAP_JSON.read_text(encoding="utf-8"))
except (FileNotFoundError, json.JSONDecodeError) as exc:
    # 模块级读取：缺失时原本抛裸 FileNotFoundError，排查体验差。
    # 换地图/新环境克隆后最常见的问题，给出生成指令（2026-08-21 深检加固）。
    raise RuntimeError(
        f"共享地图数据不可用（{MAP_JSON}）。请先运行 "
        "`python3 client/scripts/the_ville_src/convert_ville_map.py` 生成后重启后端"
    ) from exc
COLS: int = MAP_DATA["cols"]
ROWS: int = MAP_DATA["rows"]
TILE_SIZE: int = MAP_DATA["tileDim"]
# 行主序可行走网格：walkable[row][col] = True/False
WALKABLE: list[list[bool]] = MAP_DATA["walkable"]
# 玩家出生点（瓦片坐标），由转换脚本锚定在主连通区
SPAWN_COL: int = MAP_DATA["playerSpawn"]["col"]
SPAWN_ROW: int = MAP_DATA["playerSpawn"]["row"]


def is_walkable(col: int, row: int) -> bool:
    return 0 <= col < COLS and 0 <= row < ROWS and WALKABLE[row][col]


def to_tile(px: float) -> int:
    return int(px) // TILE_SIZE


def to_pixel_center(tile: int) -> int:
    return tile * TILE_SIZE + TILE_SIZE // 2


def nearest_walkable(
    col: int, row: int, max_radius: int = 12
) -> tuple[int, int] | None:
    """从 (col,row) 逐环外扩找最近可行走格；超 max_radius 环仍无则 None。

    用途：换地图后旧存档/旧 seed 坐标可能越界或落在建筑/水面——
    投射到最近可走格，而不是让居民卡在不可走点上寻路失败。
    """
    if is_walkable(col, row):
        return col, row
    for radius in range(1, max_radius + 1):
        cands: list[tuple[int, int]] = []
        for dc in range(-radius, radius + 1):  # 上下边
            cands.append((col + dc, row - radius))
            cands.append((col + dc, row + radius))
        for dr in range(-radius + 1, radius):  # 左右边（角已含）
            cands.append((col - radius, row + dr))
            cands.append((col + radius, row + dr))
        for c, r in cands:
            if is_walkable(c, r):
                return c, r
    return None

File: client/src/world/mapData.ts
// 世界静态数据：角色精灵表布局、地图 JSON v3 类型、阻挡网格几何。
//
// 2026-08-26 地图切 v3（the_ville 裁剪版，见 client/scripts/the_ville_src/）：
// - town_map_v3.json 同时是「Tiled 标准地图」（Phaser load.tilemapTiledJSON
//   原生解析多 tileset 归属 + gid 翻转标志）和「项目世界数据」（walkable/
//   sectors/locations/出生点，后端共读）——一份文件两种角色，改地图 = 改
//   转换脚本重跑，前后端自动同步。
// - 10 个渲染层按数组顺序自下而上：前 8 层永远在角色之下，Foreground
//   L1/L2（树冠/吧台等半透明前景遮挡）永远在角色之上——the_ville 是
//   娃娃房式敞开室内（数据验证：窗口内无不透明屋顶层），没有可淡出的屋顶，
//   “遮挡感”由前景层置顶实现（角色站在吧台/树冠后面被正确盖住）。
// - 阻挡不用 tile 碰撞而用预计算 walkable 网格：服务端寻路已把网格作为
//   唯一权威，前端把它按行合并成矩形静态体，物理与服务端永不打架。
// - 角色继续用 folk2 精灵表（1792×256）：8 角色 × 4 方向行 × 7 帧
//   （1 idle + 6 walk），帧 32×64。

export const FOLK_SHEET = "assets/v2/folk2.png";
export const FOLK_FRAME_W = 32;
export const FOLK_FRAME_H = 64; // 角色 1 格宽 × 2 格高
const FRAMES_PER_CHAR = 7; // 1 idle + 6 walk
const SHEET_COLS = 1792 / FOLK_FRAME_W; // 56

/** 居民 id → folk2 角色序号；玩家用 7 */
export const CHARACTER_INDEX: Record<string, number> = {
  baker_lin: 0,
  librarian_su: 1,
  florist_mo: 2,
  lao_zhou: 3,
  hong_jie: 4,
  xiao_dou: 5,
  lao_song: 6,
};
export const PLAYER_CHARACTER = 7;

export type Direction = "down" | "left" | "right" | "up";
const DIR_ROW: Record<Direction, number> = { down: 0, left: 1, right: 2, up: 3 };

/** 角色 charIndex 朝 dir 的行走帧序列（6 帧循环，不含 idle） */
export function walkFrames(charIndex: number, dir: Direction): number[] {
  const base = DIR_ROW[dir] * SHEET_COLS + charIndex * FRAMES_PER_CHAR;
  return [base + 1, base + 2, base + 3, base + 4, base + 5, base + 6];
}

/** 角色 charIndex 朝 dir 的站立帧 */
export function idleFrame(charIndex: number, dir: Direction): number {
  return DIR_ROW[dir] * SHEET_COLS + charIndex * FRAMES_PER_CHAR;
}

// ── 地图 JSON v3（the_ville 裁剪版，转换管线单一源头）────────────────────────────────────────────────

/** tileset 表项：firstgid 起 global id 连续段，columns 决定 gid→格的行列换算 */
export interface VilleTileset {
  name: string;
  firstgid: number;
  columns: number;
  tilecount: number;
  image: string;
}

/** 渲染层：data 是行主序原始 gid（含 Tiled 翻转标志位，Phaser 原生解析）。
 *  opacity/visible 是 Tiled 标准字段——Phaser v4 解析层 alpha 时直接乘
 *  curl.opacity，缺字段会得 NaN 导致整层瓦片全透明（不报错），勿删。 */
export interface VilleLayer {
  name: string;
  type: "tilelayer";
  width: number;
  height: number;
  opacity: number;
  visible: boolean;
  data: number[];
}

/** town_map_v3.json 项目自有字段（Tiled 标准字段以外部分） */
export interface VilleMapData {
  version: number;
  cols: number;
  rows: number;
  tileDim: number;
  tilesets: VilleTileset[];
  layers: VilleLayer[];
  /** 行主序可行走网格：walkable[row][col]，与服务端寻路同源 */
  walkable: boolean[][];
  playerSpawn: { col: number; row: number };
}

/** 渲染在角色之上的前景层（树冠/吧台等半透明遮挡） */
export const FOREGROUND_LAYER_NAMES = new Set(["Foreground L1", "Foreground L2"]);

/** 阻挡网格按行合并出的连续矩形（瓦片坐标，h 恒为 1——只做行内合并）。
 * v2 用逐道具碰撞体，v3 改为整网格派生：2332 个阻挡格 → ~271 个矩形，
 * 物理体数量减一个数量级且与服务端 walkable 严格一致。 */
export interface BlockedRun {
  col: number;
  row: number;
  w: number;
  h: number;
}

export function blockedRuns(walkable: boolean[][]): BlockedRun[] {
  const runs: BlockedRun[] = [];
  for (let row = 0; row < walkable.length; row++) {
    const cols = walkable[row]!;
    let start = -1;
    for (let col = 0; col <= cols.length; col++) {
      const blocked = col < cols.length && !cols[col];
      if (blocked && start < 0) start = col;
      else if (!blocked && start >= 0) {
        runs.push({ col: start, row, w: col - start, h: 1 });
        start = -1;
      }
    }
  }
  return runs;
}

/**
 * 瓦片切比雪夫距离 ≤ range：与后端 world.mapdata.to_tile + engine.CHAT_RANGE_TILES
 * 同几何。对话距离判定前后端必须一致——旧版前端用像素欧氏距离，边界处会
 * 出现"有提示却 too_far"的错位（2026-08-21 深检发现）。
 */
export function withinChebyshevTiles(
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  tileDim: number,
  rangeTiles: number,
): boolean {
  const dx = Math.abs(Math.floor(x1 / tileDim) - Math.floor(x2 / tileDim));
  const dy = Math.abs(Math.floor(y1 / tileDim) - Math.floor(y2 / tileDim));
  return Math.max(dx, dy) <= rangeTiles;
}
